src/xGadgetron/cGadgetron/include/sirf/Gadgetron/gadgetron_data_containers.h

```diff
@@ -997,34 +1074,35 @@ namespace sirf {
         void set_csm_smoothness(int s){csm_smoothness_ = s;}
 
         void calculate(CoilImagesVector& iv);
-        void calculate(const MRAcquisitionData& acq)
+        void calculate(MRAcquisitionData& acq)
         {
             CoilImagesVector ci;
             ci.calculate(acq);
             calculate(ci);
         }
 
         CFImage get_csm_as_cfimage(size_t const i) const;
+        CFImage get_csm_as_cfimage(const KSpaceSubset::TagType tag, const int offset) const;
+
 
         void get_dim(size_t const num_csm, int* dim) const
         {
             GadgetronImagesVector::get_image_dimensions(num_csm, dim);
 
         }
 
+        void forward(GadgetronImageData& img, GadgetronImageData& combined_img)const;
+        void backward(GadgetronImageData& combined_img, GadgetronImageData& img)const;
+
     protected:
 
-        //bool flag_imgs_suitable_for_csm_computation_=false;
+        void coilchannels_from_combined_image(GadgetronImageData& img, GadgetronImageData& combined_img) const;
+        void combine_images_with_coilmaps(GadgetronImageData& combined_img, GadgetronImageData& img) const;
 
-        void calculate_csm(ISMRMRD::NDArray<complex_float_t>& cm, ISMRMRD::NDArray<float>& img, ISMRMRD::NDArray<complex_float_t>& csm);
 
-        void forward(){
-            throw LocalisedException("This has not been implemented yet." , __FILE__, __LINE__);
-        }
-        void backward(){
-            throw LocalisedException("This has not been implemented yet." , __FILE__, __LINE__);
-        }
+        //bool flag_imgs_suitable_for_csm_computation_=false;
 
+        void calculate_csm(ISMRMRD::NDArray<complex_float_t>& cm, ISMRMRD::NDArray<float>& img, ISMRMRD::NDArray<complex_float_t>& csm);
```

Comment:
add `const` where you can

Reply:
same problem with const-ness as with the `ISMRMRD::Image`